### ISSUE TITLE
Inquiry on Manual Tree Annotation Workflow in CloudCompare

### ISSUE THREAD

**Issue by CiSong10** [ecker-lab/TreeLearn#19]
Dear developer,

Thanks for this powerful, easy‑to‑use repo! I’m trying to manually prepare the training data, using the steps described in your paper:

> a python script was used to iterate through all trees in the dataset. In each iteration, the central tree was loaded together with all surrounding trees in CloudCompare and all mistakes that were associated with the central tree were corrected.

After running the prediction, I see an `individual_trees/` folder containing point clouds for each tree—but without any neighboring trees. Is this what you meant by iterating through trees in the paper? These seem to contain only the points for the central tree, without the surrounding context. How did you add the surrounding context? Is it just load all trees on top of that?

Also, I’m unsure about how to edit TreeID values directly in CloudCompare. I tried using the Segment tool and then applied `Edit > Scalar Fields > Arithmetic`, first multiplying the TreeID by 0 and then adding a constant, but this feels like a workaround. Is there a more efficient method for reassigning TreeID values? If a single tree was oversegmented into multiple TreeIDs, how did you handle reassigning them to a single tree? Any tips or steps you could share on that process would be really helpful.

Thanks again for your excellent work and your time!

**Comment by CiSong10** [comment]
By the way, after training the model, is the final model saved somewhere? I used `work_dirs/epoch_100.pth` as the fine-tuned model but I am not sure if this is the best way to do it.

**Comment by jonaden94** [comment]
Dear Ci Song, 

thank you for your positive feedback on our work! Regarding your two questions:

################### Question 1
The `individual_trees/` folder contains the individual trees predicted by our method. Therefore, it is simply the output of our method (same as the result in `full_forest/`, just separated in individual point clouds), and has nothing to do with manual annotation. For manually correcting the annotations in an automatically annotated forest point cloud, we used the following workflow provided as pseudo code here:

```
for tree_id in unique_tree_ids(forest_point_cloud):
    # 1. Extract the points for current tree_id
    tree_points = extract_tree_points(forest_point_cloud, tree_id)

    # 2. Find neighboring trees (definition neighbors: minimum point distance to tree_points is below threshold)
    neighbor_ids = find_neighboring_tree_ids(forest_point_cloud, tree_points, threshold=3) # threshold of e.g. 3 meters

    # 3. Extract all relevant trees (main + neighbors)
    relevant_tree_ids = [tree_id] + neighbor_ids

    # 4. Save each tree point cloud to a separate .las file with a unique color for loading in CloudCompare
    for i, tid in enumerate(relevant_tree_ids):
        tree_pc = extract_tree_points(forest_point_cloud, tid)
        colored_pc = assign_color(tree_pc, color_from_palette(i)) # ensure different colors
        save_to_las(colored_pc, f"{tid}.las")

    # 5. Load these .las files in CloudCompare for visual inspection and correction.
        - Human manually corrects all false positive and false negative points with regard to tree_id
        - Use "cut out" and "merge" functionality in CloudCompare for this
        - After the procedure, the tree corresponding to tree_id is correctly annotated
        - Save the corrected .las files

    # 6. Load the corrected trees and update original labels in forest_point_cloud with them
    corrected_trees = load_corrected_trees()
    for tid, corrected_tree_points in corrected_trees.items():
        update_labels(forest_point_cloud, corrected_tree_points, tid) # update forest_point_cloud with corrected trees
```

Steps 1 to 4 as well as step 6 were implemented as python scripts/notebooks that were run for each unique tree id. Step 5 is the actual work done by a human annotator in CloudCompare. In each iteration, we loaded the current tree and all its neighbors as separate point clouds to CloudCompare. Assigning points from one point cloud to another one is easy then. You just have to use the "cut out" and "merge" functionality of CloudCompare. 

Of course there are some implementation details and edge cases that I did not describe here, but the general idea should be clear I hope. It should be noted that this workflow is highly time-consuming. Curating an annotated point cloud dataset requires careful planning and multiple people should be involved to avoid huge amounts of monotonous work for a single person.

################### Question 2
The best model is not saved separately. But you can select the best model by just looking at the log file. The training and validation loss are written there. In the `train.yaml` config, you can also manually change how often the model gets validated/saved by changing the validation_frequency/save_frequency argument (default is 20).


I hope this helps you
Cheers

**Comment by jonaden94** [comment]
You might also wanna contact the authors from [this paper](https://arxiv.org/abs/2309.01279). Perhaps they also have some useful tips for annotation.

**Comment by CiSong10** [comment]
Thank you very much! I will try your suggestions
